// Quantum/database.js
var mongoose = require('mongoose');
var colors = require('colors');


module.exports = {
	properties : {},
	__construct : function(properties){
		this.properties = properties;
	},	
	startConnection : function(callback){
		var objectConnection = {
			server : this.properties.server,
			base : this.properties.base,
			createSchema : this.createSchema
		};

		mongoose.connect('mongodb://'+objectConnection.server+'/'+objectConnection.base, function(error){
			if(error){
				console.log(error);
			}
			else{
				console.log(colors.green('Conectado a MongoDB \nEn el server: '
					+colors.yellow(objectConnection.server)
					+'\nA la base de datos: '+colors.yellow(objectConnection.base)));

				var confirm = objectConnection.createSchema();
				if(confirm)
					callback();
			}
		});
	},
	createSchema : function(properties){
		console.log('Creando Schema'.red);

		var Schema = mongoose.Schema;

		/* Version App */
		var appVersionSchema = new Schema({
			nameVersion : String,
			stringVersion : String,
			dateRelease : Date,
			devVersion : Boolean
		});
		module.exports = mongoose.model('appVersion', appVersionSchema);

		/* User */
		var userSchema = new Schema({
			user : String,
			name : String,
			surname : String,
			company : String,
			password : String
		});
		module.exports = mongoose.model('user', userSchema);

		return true;
	}
};
